dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java

```diff
@@ -0,0 +1,144 @@
+  private final long mPageSize =
+      Configuration.global().getBytes(PropertyKey.WORKER_PAGE_STORE_PAGE_SIZE);
+
+  public DoraMetaManager(
+      PagedDoraWorker doraWorker, DoraMetaStore metaStore, CacheManager cacheManger,
+      UnderFileSystem ufs, LockPool<String> pathLocks) {
+    mMetastore = metaStore;
+    mCacheManager = cacheManger;
+    mDoraWorker = doraWorker;
+    mUfs = ufs;
+    mPathLocks = pathLocks;
+  }
+
+  /**
+   * Gets file meta from UFS.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> getFromUfs(String path) throws IOException {
+    try {
+      UfsStatus status = mUfs.getStatus(path);
+      DoraMeta.FileStatus fs = mDoraWorker.buildFileStatusFromUfsStatus(status, path);
+      return Optional.ofNullable(fs);
+    } catch (FileNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Gets file meta from UFS and loads it into metastore if exists.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> loadFromUfs(String path) throws IOException {
+    Optional<FileStatus> fileStatus = getFromUfs(path);
+    try (var lockResource = mPathLocks.get(path, LockMode.WRITE)) {
+      fileStatus.ifPresent(status -> put(path, status));
+      return fileStatus;
+    }
+  }
+
+  /**
+   * Gets file meta from the metastore.
+   * @param path the full ufs path
+   * @return the file status, or empty optional if not found
+   */
+  public Optional<FileStatus> getFromMetaStore(String path) {
+    try (var lockResource = mPathLocks.get(path, LockMode.READ)) {
+      return mMetastore.getDoraMeta(path);
+    }
+  }
+
+  /**
+   * Puts meta of a file into the metastore, and invalidates the file data cache.
+   * @param path the full ufs path
+   * @param meta the file meta
+   */
+  public void put(String path, FileStatus meta) {
+    try (var lockResource = mPathLocks.get(path, LockMode.WRITE)) {
+      Optional<FileStatus> status = mMetastore.getDoraMeta(path);
+      if (status.isPresent() && !status.get().getFileInfo().getFolder()
+          && status.get().getFileInfo().getLength() > 0) {
+        Fingerprint originalFingerprint =
+            Fingerprint.parse(status.get().getFileInfo().getUfsFingerprint());
+        if (originalFingerprint == null) {
+          originalFingerprint = Fingerprint.INVALID_FINGERPRINT;
+        }
+        Fingerprint newFingerprint =
+            Fingerprint.parse(meta.getFileInfo().getUfsFingerprint());
+        if (newFingerprint == null) {
+          newFingerprint = Fingerprint.INVALID_FINGERPRINT;
+        }
+        boolean contentMatched = originalFingerprint.isValid()
+            && newFingerprint.isValid()
+            && originalFingerprint.matchContent(newFingerprint);
+        if (!contentMatched) {
+          invalidateCachedFile(path, status.get().getFileInfo().getLength());
```

Comment:
i discussed with @huanghua78 on this
we feel invalidate the cache file synchronously might create some performance issues and ideally it should be async. @beinan also proposed an asynchronous update that can also provides consistency. Basically record the last modified time of the metadata and for each cached page we also record the loaded time. The client will attaches the last modified time of the metadata when it reads the data and the server can rejects the client if the data becomes stale. 

But we guess in the short term we should still go with the sync approach because:
1. for the time being, update metadata is rare 
2. the async implementation is not trivial and needs a more formal implementation 
3. letting the client carry the timestamp is not easy, it might need some client side code change too (e.g. a hashmap for the timestamp)